proposals/0010-vk-buffer-ref.md

```diff
@@ -0,0 +1,270 @@
+* There is no equivalent to the `*` operator that will dereference the entire struct.
+* There is no conversion from uint64_t to vk:buffer_ptr.
+* There is no explicit pointer arithmetic. All addressing is implicitly done using the `.` pointer, or indexing an array in the struct T.
+* The comparison operators == and != are not supported for buffer pointers. Comparison to null pointer should be done by first converting to uint64_t.
+
+Most of these restrictions are there for safety. They minimize the possibility of getting an invalid pointer. If the `.` operator is used on a null pointer, the behaviour is undefined.
+
+When used as a member in a buffer, vk::buffer_ptr can be used to pass physical buffer addresses into a shader, and address and access buffer space with logical addressing, which allows tools such as spirv-opt, spirv-reflect and renderdoc to be able to better work with these shaders.
+
+For example, here is a shader using vk::buffer_ptr to do the same thing as the shader above using vk::RawBufferLoad. Note the natural, logical syntax of the reference:
+
+```c++
+
+struct Globals_s
+{
+      float4 g_vSomeConstantA;
+      float4 g_vTestFloat4;
+      float4 g_vSomeConstantB;
+};
+
+typedef vk::buffer_ptr<Globals_s> Globals_p;
+
+struct TestPushConstant_t
+{
+      Globals_p m_nBufferDeviceAddress;
+};
+
+[[vk::push_constant]] TestPushConstant_t g_PushConstants;
+
+float4 MainPs(void) : SV_Target0
+{
+      float4 vTest = g_PushConstants.m_nBufferDeviceAddress.g_vTestFloat4;
+      return vTest;
+}
+
+```
+
+In SPIR-V, Globals_t would be a pointer to the physical buffer storage class. The type of the push constant would contain one of those pointers. The SPIR-V for this shader can be seen in Appendix B. Note the logical context of the declaration and addressing of underlying struct Globals_t including Offset decorations all Globals_t members.
+
+## Linked Lists and Local Variables
+
+vk::buffer_ptr can be used to program a linked list of identical buffers:
+
+```c++
+
+// Forward declaration
+typedef struct block_s block_t;
+typedef vk::buffer_ptr<block_t> block_p;
+
+struct block_s
+{
+      float4 x;
+      block_p next;
+};
+
+struct TestPushConstant_t
+{
+      block_p root;
+};
+
+[[vk::push_constant]] TestPushConstant_t g_PushConstants;
+
+float4 MainPs(void) : SV_Target0
+{
+      block_p g_p = g_PushConstants.root;
+      g_p = g_p.next;
+      if (uint64_t(g_p) == 0) return float4(0.0,0.0,0.0,0.0);
+      return g_p.x
+}
+
+```
+
+Note also the ability to create local variables of type vk::buffer_ptr such as g_p which can be read, written and dereferenced.
+
+## Design Details
+
+### Differences from C++ Pointers
+
+vk::buffer_ptr is different from a C++ pointer in that a selection “.” operation can and must be applied to a buffer pointer to de-reference it. 
```

Comment:
I think this conflicts with the future directions of HLSL, so it would be better to add a more forward looking syntax.

Reply:
OK